lib/rules.js

```diff
@@ -353,6 +353,267 @@ export const GROUP_TERMS_BY_ROOT = defineRule(
+
+// e.g. (2x)^2 - (3y)^2 -> (2x + 3y)(2x - 3y)
+export const FACTOR_DIFFERENCE_OF_SQUARES =
+    defineRuleString('#a^2 - #b^2', '(#a + #b)(#a - #b)')
+
+
+export const nthRoot = (num, root = 2, precision = 12) => {
+    // e.g 2^-3 = 1/(2^3)
+    const inv = root < 0
+    if (inv) {
+        root = -root
+    }
+
+    if (root === 0) {
+        throw new Error('Root must be non-zero')
+    }
+    if (num < 0 && (Math.abs(root) % 2 !== 1)) {
+        throw new Error('Root must be odd when a is negative.')
+    }
+
+    // Edge cases zero and infinity.
+    // e.g 0^3 = 0, 0^-3 = Infinity
+    if (num === 0) {
+        return inv ? Infinity : 0
+    }
+
+    if (num === 1) {
+        return 1
+    }
+
+    if (!isFinite(num)) {
+        return inv ? 0 : num
+    }
+
+    // Source: https://rosettacode.org/wiki/Nth_root#JavaScript
+    const n = root
+    const prec = precision
+
+    let x = 1 // Initial guess
+    for (let i = 0 ; i < prec ; i++) {
+        x = 1 / n * ((n - 1) * x + (num / Math.pow(x, n - 1)))
+    }
+
+    return inv ? 1 / x : x
+}
+
+// TODO: handle fractional coefficients
+// TODO: isPolynomialTerm should handle fractional coeffs
+
+// e.g. 4x^2 + 12x + 9 -> (2x + 3)^2
+export const FACTOR_PERFECT_SQUARE_TRINOMIALS = defineRule(
+    (node) => {
+        const isFactorable = canApplyRule(FACTOR_SUM_PRODUCT_RULE, node)
+        if (isFactorable) {
+            const result = applyRule(FACTOR_SUM_PRODUCT_RULE, node)
+            return print(result.args[0]) === print(result.args[1])
+                ? {node} : null
+        }
+    },
+
+    (node) => {
+        const result = applyRule(FACTOR_SUM_PRODUCT_RULE, node)
+        return build.apply('pow', [result.args[0], build.numberNode(2)])
+    }
+)
+
+function getFactors(num){
+    var factors = []
+
+    for(var i = 1; i <= Math.sqrt(num); i++){
+        if(Number.isInteger(num/i)){
+            factors.push([i, num/i])
+        }
+    }
+
+    return factors
+}
+
+// TODO: handle case when exponent is a polynomial (x^b^4)
+// e.g. 12x^2 + 17x + 6 -> (3 x^1 + 2) (4 x^1 + 3)
+export const FACTOR_SUM_PRODUCT_RULE = defineRule(
+    (node) => {
+        if (query.isAdd(node) && node.args.length === 3) {
+            const [firstTerm, secondTerm, thirdTerm] = node.args
+
+            // First two terms should be polynomials
+            // Third term is either number or polynomial
+            const isTrinomial =
+                  isPolynomialTerm(firstTerm)
+                  && isPolynomialTerm(secondTerm)
+                  && (query.isNumber(thirdTerm) || isPolynomialTerm(thirdTerm))
+
+            if (isTrinomial) {
+                let isFactorable = false
+                const {constants, coefficientMap} = getCoefficientsAndConstants(node)
+                const variables = Object.keys(coefficientMap).map(key => parse(key))
+                const coefficients = Object.keys(coefficientMap).map(key => {return coefficientMap[key]})
```

Comment:
```
const coeffs = Object.keys(coeffMap).map(key => coeffMap[key][0])
```
This keeps it to a single line and you don't have to do `coeffs[1][0]` later you can just do `coeffs[1]`.